java/ql/lib/semmle/code/java/Modifier.qll

```diff
@@ -46,8 +46,11 @@ abstract class Modifiable extends Element {
   // TODO: `isSealed()` conflicts with `ClassOrInterface.isSealed()`. What name do we want to use here?
   predicate isSealedKotlin() { this.hasModifier("sealed") }
 
-  /** Holds if this element has a `public` modifier or is implicitly public. */
-  predicate isPublic() { this.hasModifier("public") }
+  /**
+   * Holds if this element has a `public` modifier or is implicitly public.
+   * Kotlin `internal` members, which are `public` in JVM Bytecode, are not considered `public`.
+   */
+  predicate isPublic() { this.hasModifier("public") and not this.isInternal() }
```

Comment:
I wanted to introduce as little change as possible. `hasModifier` has a comment on top of it:
```
  /**
   * Holds if this element has modifier `m`.
   *
   * For most purposes, the more specialized predicates `isAbstract`, `isPublic`, etc.
   * should be used.
   *
   * Both this method and those specialized predicates take implicit modifiers into account.
   * For instance, non-default instance methods in interfaces are implicitly
   * abstract, so `isAbstract()` will hold for them even if `hasModifier("abstract")`
   * does not.
   */
  predicate hasModifier(string m) { modifiers(this.getAModifier(), m) }
```
So in my view it's okay to return the `public` modifier even if `isPublic` doesn't hold. That feels similar to the `abstract` case.

I think the best option would be making `hasModifier` private, but it's used heavily for concatenating all modifiers. (For all other purposes we anyways should go with the specialized predicates, which in some case are overridden (for example `Method::isPublic`).)

What do you think?

Reply:
My proposal is that codeql treats internal predicates as simply internal. The only reason that we sometimes get a `public` modifier in the database is that the Java extractor isn't able to tell that they are internal, and it isn't easy to patch that up before the database is created, and so I suggest that we patch it at the lowest level of QL, i.e. wrapping the relevant table and only using the wrapper elsewhere.

As far as queries are concerned, there should therefore be no public modifier, whether you ask `isPublic` or `hasModifier`. This isn't a case of it being implicit, it simply shouldn't exist.

An alternative approach would be to always extract `internal` entities as also being `public` in the Kotlin extractor, so entities would always have both modifiers consistently. This would presumably be consistent with Java's world view, but I think that it would be more confusing on balance.

Reply:
I see what you mean. How do you like https://github.com/github/codeql/pull/10221/commits/e66d2dddb68edb10fb25eb0f09b8bf94ad35c617?